### PR TITLE
feat(ext/ffi): Implement `UnsafePointer` and `UnsafePointerView` API

### LINKED ISSUE #12809: feature request: support for null value in FFI buffer argument
There's no way to pass a null pointer to a buffer argument in FFI right now. It would be great to support passing JS `null` value corresponding to a `nullptr` in FFI. I'm currently trying to use WinAPI which (of course) makes use of those.

I'd like to try work on this

### PR DESCRIPTION
**Pull request by eliassjogreen**
This pr adds support for passing, returning and reading pointers as proposed by the `UnsafePointer` and `UnsafePointerView` api (#12824).

Closes #12809, #12456, #12824 and #12752

### CODE REVIEW

**Comment by eliassjogreen** [comment]
This pr could very easily implement `UnsafePointer` as proposed in #12824 if that is what we want. It would most likely make working with pointers way easier imo

**Comment by bnoordhuis** [comment]
I'm not sure using f64's is sound. V8 tends to canonicalize NaNs - maybe Infs and subnormals too, but definitely NaNs - which means there's a subset of the 64 bit space that's going to get corrupted.

You could _maybe_ get away with it if it's only NaNs and Infs<sup>1</sup> but definitely not when it's subnormals too. An example of a subnormal is `0000 0000 0000 0001` (`1u64` interpreted as f64.)

tl;dr Don't rely on V8 leaving the top 12 bits untouched, use BigInts instead.

<sup>1</sup> ARM64 and x86_64 have up to 52 address lines, which is exactly the f64 significand's size. NaNs and Infs are encoded in the exponent, not the significand.

**Comment by eliassjogreen** [comment]
Would returning and passing pointers as a tuple of the low and high part of a `u64` be preferred or should we wait for `serde_v8` to support `bigint` (de)serialization?

**Comment by eliassjogreen** [comment]
@bnoordhuis internally pointers are now represented as the high and low bytes of an `u64` (`(u32, u32)`) which should avoid the issues with transmuting pointers as `f64`. I also implemented the `UnsafePointer` API to make it easier to work with pointers although it should probably be looked through for my sanity and to make sure it all works as expected, cc @DjDeveloperr

File: ext/ffi/00_ffi.js [hunk 1]
```diff
@@ -6,7 +6,52 @@
...
+  function unpackPointer([hi, lo]) {
+    return BigInt(hi) << 32n | BigInt(lo);
+  }
+
+  function packPointer(value) {
+    return [Number(value >> 32n), Number(value & 0xFFFFFFFFn)];
+  }
+
+  class UnsafePointer {
+    value;
```
Comment (DjDeveloperr) [inline ext/ffi/00_ffi.js:23]: Could `valueOf` method be implemented as well?

File: ext/ffi/lib.rs [hunk 2]
```diff
@@ -108,6 +112,9 @@ pub fn init<P: FfiPermissions + 'static>(unstable: bool) -> Extension {
       ("op_ffi_load", op_sync(op_ffi_load::<P>)),
       ("op_ffi_call", op_sync(op_ffi_call)),
       ("op_ffi_call_nonblocking", op_async(op_ffi_call_nonblocking)),
+      ("op_ffi_ptr_of", op_sync(op_ffi_ptr_of)),
```
Comment (DjDeveloperr) [inline ext/ffi/lib.rs:115]: Unsafe Pointer class' methods like static `of`, `read`, `readCString` should require `allow-ffi` permission IMO

**Review by DjDeveloperr — COMMENTED** [review]
LGTM, got a few comments

File: ext/ffi/00_ffi.js (hunk 1 above)
Comment (eliassjogreen) [reply ext/ffi/00_ffi.js:23]: Hmm, yeah but im undecided if we want pointer values to be immutable because the `UnsafePointer` constructor allows the user to create new pointers from an arbitary value so why not then allow the value property to be mutable and just leave the value property exposed.

File: ext/ffi/00_ffi.js [hunk 3]
```diff
@@ -6,7 +6,52 @@
...
+
+    constructor(value) {
+      this.value = value;
+    }
+
+    static null() {
```
[the remark this answers is not in the document]
Comment (eliassjogreen) [reply ext/ffi/00_ffi.js:29]: I agree with this, will fix

File: ext/ffi/00_ffi.js (hunk 1 above)
Comment (DjDeveloperr) [reply ext/ffi/00_ffi.js:23]: Ah, no I didn't mean making value immutable, but just adding `valueOf` method too, it would allow things such as `new UnsafePointer(1n) + 2n`

Comment (eliassjogreen) [reply ext/ffi/00_ffi.js:23]: ah ok, didn't know about that but sure!

File: ext/ffi/lib.rs (hunk 2 above)
Comment (eliassjogreen) [reply ext/ffi/lib.rs:115]: Hmm, how would I implement this? `FfiPermissions::check` requires a `path` component, should i expose and use `FfiDescriptor::query` or how would i check if `allow-ffi` is enabled without checking any specific path?

File: cli/dts/lib.deno.unstable.d.ts [hunk 4]
```diff
@@ -117,16 +117,59 @@ declare namespace Deno {
...
+     * Return the direct memory pointer to the typed array in memory
+     */
+    static of(typedArray: TypedArray): UnsafePointer;
+
+    /**
+     * Reads memory at the pointer into a typed array.
+     * Length is determined from the typed array's `byteLength`.
+     * Also takes optional offset.
+     */
+    read(destination: TypedArray, offset?: bigint | number): void;
```
Comment (rracariu) [inline cli/dts/lib.deno.unstable.d.ts:160]: My understanding is that this would copy data, would be nice to have an view based approach over a said pointer into native memory that uses https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView as a way to read typed information from the raw memory the pointer points to.

Comment (eliassjogreen) [reply cli/dts/lib.deno.unstable.d.ts:160]: Perhaps a read only (kinda dangerous writing to arbitrary pointers, but maybe...) copy of the `DataView` API? Although there is a few questions i have with it:
1. It would have to work with `bigint` offsets instead of `numbers`
2. Why would you want to create more than one instance of this new `DataView` copy as they all address the same memory? Or should it be an extension to the `UnsafePointer` API?
3. The `read` method should probably also be renamed `copy` as that more accurately reflects what it does, or maybe we should even add a `getArrayBuffer(byteLength: bigint | number, offset?: bigint | number): ArrayBuffer` (or `Uint8Array` maybe?)

Comment (rracariu) [reply cli/dts/lib.deno.unstable.d.ts:160]: The new structure should mimic the `DataView` API, work with `bigint` as you said.

I think having 2 methods on the `UnsafePointer` will be nice:
- `UnsafePointer::toReadonlyView(offset:bignum, size: bignum): ReadOnlyPointerView`
- `UnsafePointer::toWritableView(offset:bignum, size: bignum): WritablePointerView`
Where `WritablePointerView` would extend `ReadOnlyPointerView` and allow write access to the underlying memory. This is just a proposal so naming and form could be definitely be improved...

I agree that the `read` method should be named to `copy` as it as clear as possible about the intent.
Maybe create a new PR with this new `View` APIs so it is easier to review and make this one mergeable quicker?

Comment (DjDeveloperr) [reply cli/dts/lib.deno.unstable.d.ts:160]: > The `read` method should probably also be renamed `copy` as that more accurately reflects what it does, or maybe we should even add a `getArrayBuffer(byteLength: bigint | number, offset?: bigint | number): ArrayBuffer` (or `Uint8Array` maybe?)

I had it to be named `read` in the proposal as it seemed consistent with `Deno.Reader#read`. But rename to `copy` would make sense, I agree.

Comment (eliassjogreen) [reply cli/dts/lib.deno.unstable.d.ts:160]: > Maybe create a new PR with this new `View` APIs so it is easier to review and make this one mergeable quicker?

Ideally yes, but should this PR then leave out the `read` and `readCString` methods in favour of that future pr?

**Comment by eliassjogreen** [comment]
This PR is ready for review now!

File: ext/ffi/lib.rs [hunk 5]
```diff
@@ -267,6 +296,21 @@ fn value_as_f64(value: Value) -> f64 {
     .expect("Expected ffi arg value to be a float")
 }
 
+#[derive(Serialize, Deserialize, Debug)]
+struct U32x2(u32, u32);
+
+impl From<u64> for U32x2 {
+  fn from(value: u64) -> Self {
+    U32x2((value >> 32) as u32, value as u32)
```
Comment (bnoordhuis) [inline ext/ffi/lib.rs:304]:
```suggestion
    Self((value >> 32) as u32, value as u32)
```
For consistency?

File: ext/ffi/lib.rs [hunk 6]
```diff
@@ -507,6 +560,177 @@ async fn op_ffi_call_nonblocking(
...
+  _: (),
+) -> Result<(), AnyError>
+where
+  FP: FfiPermissions + 'static,
+{
+  let permissions = state.borrow_mut::<FP>();
+  permissions.check(None)?;
+
+  let src = u64::from(src) as *const u8;
+  unsafe { ptr::copy(src, dst.as_mut_ptr(), len) };
```
Comment (bnoordhuis) [inline ext/ffi/lib.rs:589]: Suggestion: return an error if `dst.len() < len`?

File: ext/ffi/lib.rs [hunk 7]
```diff
@@ -507,6 +560,177 @@ async fn op_ffi_call_nonblocking(
...
+  ptr: U32x2,
+  _: (),
+) -> Result<u16, AnyError>
+where
+  FP: FfiPermissions + 'static,
+{
+  let permissions = state.borrow_mut::<FP>();
+  permissions.check(None)?;
+
+  Ok(unsafe { *(u64::from(ptr) as *const u16) })
```
Comment (bnoordhuis) [inline ext/ffi/lib.rs:647]: This won't work on architectures where unaligned loads trap. Use `std::ptr::read_unaligned()` instead.

**Review by bnoordhuis — COMMENTED** [review]
Nice work. Mostly LGTM.

File: ext/ffi/lib.rs (hunk 6 above)
Comment (eliassjogreen) [reply ext/ffi/lib.rs]: Actually this won't ever be a problem as the one place where it is called is:
```js
core.opSync("op_ffi_buf_copy_into", [
  packU64(this.pointer.value + BigInt(offset)),
  destination,
  destination.byteLength,
]);
```
and here the length is always the length of the `dst` variable, but implemented it anyways for good measure.

**Comment by brandonros** [comment]
What next release can we expect this to land in? v1.17.0?

**Comment by DjDeveloperr** [comment]
Yes, it's going to be in 1.17.0 release which is a bit delayed it seems

**Comment by lucacasonato** [comment]
@DjDeveloperr It is not delayed. 1.17.0 will release tomorrow as documented in the release schedule: https://deno.land/manual@v1.16.4/contributing/release_schedule. I have updated the milestone accordingly now.
